### ISSUE TITLE
CA1823 incorrectly flags const used as a fixed buffer length

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#37593]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.4

#### Diagnostic ID

[CA1823](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1823-avoid-unused-private-fields?view=vs-2019)

#### Repro steps

```C#
using System.Runtime.InteropServices;

class Program
{
    static void Main() { }

    private const int SomeLength = 16;

    [StructLayout(LayoutKind.Sequential)]
    internal unsafe struct SomeStruct
    {
        internal fixed byte FixedBuffer[SomeLength];
    }
}
```

#### Expected behavior
No warnings.

#### Actual behavior
The CA1823 analyzer flags the SomeLength const as being unused:
```
Program.cs(7,23,7,33): warning CA1823: Unused field 'SomeLength'.
```
even though obviously it is being used.

**Comment by stephentoub** [comment]
> Even Find All References also does not seem to find this reference to SomeLength

Yeah, and the Rename support doesn't fix the reference, IDE0051 flags it as unused, Go to Definition on the reference fails, etc.

**Comment by CyrusNajmabadi** [comment]
QuiclInfo doesn't work either.  So this is likely a deeper SemanticModel issue, not just an IOp issue.

**Comment by mavasani** [comment]
@CyrusNajmabadi is correct - this is a deeper issue in the compiler layer. Moved to Area-Compilers.

**Comment by sharwell** [comment]
The underlying semantic model issue is likely #29224

**Comment by gwr** [comment]
FYI, there's an instance of this warning in PR https://github.com/dotnet/runtime/pull/105403
src/libraries/Common/src/Interop/SunOS/procfs/Interop.ProcFs.Definitions.cs
Would be nice to remove that pragma once this is fixed.
Let me know if a separate issues is wanted for that.

**Comment by Rekkonnect** [comment]
I've left comments in this one #75113: https://github.com/dotnet/roslyn/issues/75113#issuecomment-2409133546

Quoting:
> The fix is non-trivial because of binding issues. It seems that we need to implement a brand new bound node, or include field buffer size info in the bound field declaration to represent the fixed field initializer, which is currently not supported in the bound node model. Since we cannot bind the node to an appropriate bound node, we are also blocked by bound node walkers like NullableWalker and others, and we therefore cannot evaluate the constituting expressions.
